health_check.go

```diff
@@ -20,48 +24,103 @@ type HealthCheck struct {
+		time.Sleep(10 * time.Millisecond)
+	}
 
-		// Set defaults
-		if check.RequestMethod == "" {
-			check.RequestMethod = "GET"
-		}
-		if check.ResponseStatusCodeEquals == 0 {
-			check.ResponseStatusCodeEquals = 200
-		}
+	waitGroup.Wait()
+	close(issueChan)
 
-		// Use retryablehttp to prevent false positives.
-		req, err := retryablehttp.NewRequest(check.RequestMethod, check.RequestURL, []byte(check.RequestBody))
-		if err != nil {
-			log.Printf("Health check %s: %s", check.RequestURL, err)
-			issues = append(issues, issueEntry{warning, fmt.Sprintf("%s: invalid health check", check.RequestURL)})
-			continue
-		}
-		for key, value := range check.RequestHeaders {
-			req.Header.Set(key, value)
+	for issue := range issueChan {
+		if issue != nil {
+			issues = append(issues, *issue)
 		}
+	}
+	return
+}
+
+func runHealthCheck(check HealthCheck) *issueEntry {
+	log.Printf(" checking HTTP endpoint %s", check.RequestURL)
+
+	// Set defaults
+	if check.RequestMethod == "" {
+		check.RequestMethod = "GET"
+	}
+	if check.ResponseStatusCodeEquals == 0 {
+		check.ResponseStatusCodeEquals = 200
+	}
+
+	// Use retryablehttp to prevent false positives.
+	req, err := retryablehttp.NewRequest(check.RequestMethod, check.RequestURL, []byte(check.RequestBody))
+	if err != nil {
+		log.Printf("Health check %s: %s", check.RequestURL, err)
+		return &issueEntry{warning, fmt.Sprintf("%s: invalid health check", check.RequestURL)}
+	}
+	for key, value := range check.RequestHeaders {
+		req.Header.Set(key, value)
+	}
+
+	var intermediateIssue *issueEntry
 
-		resp, err := retryablehttp.NewClient().Do(req)
-		if err != nil {
-			issues = append(issues, issueEntry{danger, fmt.Sprintf("%s: cannot be reached", check.RequestURL)})
-			continue
+	client := retryablehttp.NewClient()
+	client.HTTPClient.Timeout = 3 * time.Second
+	client.CheckRetry = func(ctx context.Context, resp *http.Response, respErr error) (bool, error) {
+		retry, err2 := retryablehttp.DefaultRetryPolicy(ctx, resp, respErr)
+		if !retry {
+			return false, err2
 		}
-		if resp.StatusCode != check.ResponseStatusCodeEquals {
-			issues = append(issues, issueEntry{danger, fmt.Sprintf("%s: received unexpected status code %d", check.RequestURL, resp.StatusCode)})
-			continue
+		if intermediateIssue == nil {
+			intermediateIssue = generateHealthCheckIssueEntry(check, resp, respErr)
 		}
+		return true, err2
+	}
 
-		for key, value := range check.ResponseHeaderContains {
-			if resp.Header.Get(key) != value {
-				issues = append(issues, issueEntry{danger, fmt.Sprintf("%s: expected response header \"%s: %s\" could not be found", check.RequestURL, key, value)})
-			}
+	resp, err := client.Do(req)
+	issue := generateHealthCheckIssueEntry(check, resp, err)
+	if issue != nil {
+		return issue
+	}
+
+	// Generate warning if health check was unstable.
+	if intermediateIssue != nil {
+		return &issueEntry{
+			issueType: warning,
+			message:   fmt.Sprintf("Unstable health check: %s", intermediateIssue.message),
```

Comment:
Nitpick: since `intermediateIssue` and the return value have the same type, it would I think be better to overwrite the fields of `intermediateIssue` to what they need to be and then return `intermediateIssue`. Otherwise, if we would later add a field to the `issueEntry` struct, this code is going to miss it.